reader_concurrency_semaphore.cc

```diff
@@ -653,8 +653,16 @@ future<> reader_concurrency_semaphore::execution_loop() noexcept {
     }
 }
 
-void reader_concurrency_semaphore::consume(const resources& r) noexcept {
+void reader_concurrency_semaphore::consume(resources r) noexcept {
     _resources -= r;
+    //TODO: make this a ctor parameter and expose as a (live-update) config option.
+    const size_t limit_multiply_threshold = 3;
+    if (!is_unlimited() && _resources.memory < 0 && -_resources.memory > _initial_resources.memory * (limit_multiply_threshold - 1)) {
+        auto& largest_contributor = *std::max_element(_permit_list.begin(), _permit_list.end(), [] (const reader_permit::impl& a, const reader_permit::impl& b) {
+            return a.resources().memory < b.resources().memory;
+        });
+        largest_contributor.get_abort_source().request_abort_ex(std::make_exception_ptr(std::runtime_error("OOM killed")));
+    }
```

Comment:
We can go crazy about the heuristics here.

1. Should we preserve the oldest query? So it can make progress. If we assume the user will retry failing queries, then making sure that eventually a query will become the oldest one, and then we can try to make it succeed.
2. I think we should have two threshold: a lower one for killing almost everything (but preserving one query so that the system as a whole makes progress) and a higher one (killing possibly everything, but protecting developers from having to debug core dumps)

Reply:
We don't need to kill more than one read here. If the memory consumption keeps ballooning we will soon get above the limit and kill another read.

Reply:
Doesn't it take time for the kill to reclaim memory? We may kill any queries before we reclaim something.

And, I do think we need a different threshold for the next-to-last query and the last query. But we can wait with that.

Reply:
Right, since aborting is cooperative, it might take some time for one of the sub-readers to step on it.

Reply:
After thinking about this a little bit more, a few problems occurred to me:
* it might be a while until the aborted reader steps on the abort source, all the while this condition can be triggered many times
* following from the above, we need to make sure here we don't try to abort the same reader(s) twice
* what to do if the readers we want to abort just don't want to die (they are very slow)

There is a more simple solution here: just throw `std::bad_alloc`. Yes, it is quite brutish and not fair, whoever happens to make the an allocation at the wrong time gets the failure, not necessarily the reader who will release the most memory if killed. On the other hand throwing here is immediate: the reader stepping on it will immediately start to unwind, providing release. We are also guaranteed to punish a reader that is actively allocating. If you think about it we use the same "algorithm" (hammer really) when we run out of memory. We don't try to be clever, we just use the bad-alloc hammer on however is unlucky enough to have requested one byte too much.